documentation_builder/metadata.ipynb

```diff
@@ -0,0 +1,548 @@
+{
+ "nbformat": 4,
+ "nbformat_minor": 0,
+ "metadata": {
+  "kernelspec": {
+   "display_name": "Python 3",
+   "language": "python",
+   "name": "python3"
+  },
+  "language_info": {
+   "codemirror_mode": {
+    "name": "ipython",
+    "version": 3
+   },
+   "file_extension": ".py",
+   "mimetype": "text/x-python",
+   "name": "python",
+   "nbconvert_exporter": "python",
+   "pygments_lexer": "ipython3",
+   "version": "3.6.0"
+  },
+  "colab": {
+   "name": "metadata.ipynb",
+   "provenance": []
+  }
+ },
+ "cells": [
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "id": "bVgYSbcPLbwU",
+    "colab_type": "text"
+   },
+   "source": [
+    "# Metadata\n",
+    "\n",
+    "Metadata in the form of annotations, provenance or additional key:value pairs is an important asset for a model.\n",
+    "Such data can be handled via the `annotation` field on cobra objects"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "id": "Rv-QKmAPM8PY",
+    "colab_type": "text",
+    "pycharm": {
+     "name": "#%% md\n"
+    }
+   },
+   "source": [
+    "# History\n",
+    "\n",
+    "History attribute, present in the `annotation` attribute of a COBRA component, stores the history of the component like creators, created date and modified date of that component."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "metadata": {
+    "pycharm": {
+     "name": "#%%\n"
+    },
+    "id": "EsBAtYeHLbwc",
+    "colab_type": "code",
+    "colab": {}
+   },
+   "source": [
+    "from cobra.core import Metabolite\n",
+    "from cobra.core.metadata import History, Creator, HistoryDatetime\n",
+    "\n",
+    "metabolite = Metabolite(id=\"glc\", name=\"D-glucose\")\n"
+   ],
+   "execution_count": null,
+   "outputs": []
+  },
+  {
+   "cell_type": "markdown",
+   "source": [
+    "The dates have to be objects of `HistoryDateTime` class and creators that of `Creator` class. We can add then directly via constructor."
+   ],
+   "metadata": {
+    "collapsed": false
+   }
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "outputs": [],
+   "source": [
+    "history = History(\n",
```

Comment:
This looks very clunky with all the constructors, type casting. This would be much more user-friendly with a simpler API.

Reply:
I'm thinking NamedTuple, which will unfortunately be immutable. Python 3.7 has dataclass, which might be even better, but would force cobrapy to use Python 3.7. What do you think? NamedTuple or dataclass?